Add `Payload` module for serializing and deserializing

jcredding opened the pull request:
This adds a `Payload` module that is responsible for serializing
jobs and deserializing serialized payloads. This doesn't change
any functionality; its just an organization change. It centralizes
all the serialization and deserialization for jobs into one
location and is setup for changing events to inherit from job.

@kellyredding - Ready for review.

kellyredding commented:
@jcredding ok, bear with me here.  I saw this implementation and my first thought was "awesome" and "this looks great".  Then I kept looking at it and how it was used in the rest of the code and something started to bother me: how these functions are named.

You know me - I put on my "dummy hat" and started to get confused about what is doing what.  Take the `deserialize` function in the `Payload` module.  There are *3 different* uses of the word "deserialize" all jux'd together.  That bothers me.

Anyway, I have a couple of ideas to address this:

1) I'd like to propose renaming the `Qs` namespace `serialize` and `deserialize` methods  to `encode` and `decode` instead.  I know this is kind of a big change.  I would assume we'd rename the config to be `encoder` and `decoder` as well and there would be other changes too (maybe changing some local vars?).

2) I'd like to propose an alternate implementation for the functions in the `Payload` module.  It is very similar to you implementation (which is sound as is) but just changes some names:

```ruby
require 'qs'
require 'qs/job'

module Qs

  module Payload

    # I like this b/c it reads:
    # "Give me the payload job from the decoded serialized payload"
    def self.deserialize(serialized_payload)
      self.job(Qs.decode(serialized_payload))
    end

    # I like this b/c it reads:
    # "Give me the payload job for this payload hash"
    def self.job(payload_hash)
      Qs::Job.new(payload_hash['name'], payload_hash['params'], {
        :type       => payload_hash['type'],
        :created_at => Time.at(payload_hash['created_at'].to_i)
      })
    end

    # I like this b/c it reads:
    # "Give me the encoded payload hash of the job"
    def self.serialize(job)
      Qs.encode(self.hash(job))
    end

    # I like this b/c it reads:
    # "Give me the payload hash for this job"
    def self.hash(job)
      { 'type'       => job.payload_type.to_s,
        'name'       => job.name.to_s,
        'params'     => StringifyParams.new(job.params),
        'created_at' => job.created_at.to_i
      }
    end

    # ...

  end

end
```

I propose and would like to see what the effect of these changes has on the implementation.  Besides the comments above about what I like, I also like that:

* there is ONE "deserialize/serialize" method
* "encode/decode" seems to better describe what that function is doing relative to Qs.  By that I'm proposing that "deserialize/serialize" relative to Qs involves job objects not just hashes (it is more than what "deserialize/serialize" relative to json is).
* the payload "job" and "hash" methods describe what you are getting, not what arguments they are taking
* the payload "serialize" and "deserialize" methods are very symmetrical in this implementation

Anyway, long freaking thought/idea here.  What do you think?  I know it is a bigger change (sorry about that).  I think I prefer this proposal, but gotta know your thoughts to be certain.  Thanks man.

kellyredding commented:
@jcredding I'm going to make some specific line comments to be clear about what I don't care for in the current implementation, FYI.

kellyredding commented on lib/qs/payload.rb:
```diff
@@ -0,0 +1,47 @@
+
+    def self.deserialize(serialized_payload)
+      self.deserialize_hash(Qs.deserialize(serialized_payload))
+    end
```
@jcredding the jux'ing of many "deserialize"s here bothers me.

kellyredding commented on lib/qs/payload.rb:
```diff
@@ -0,0 +1,47 @@
+        'params'     => StringifyParams.new(job.params),
+        'created_at' => job.created_at.to_i
+      })
+    end
```
@jcredding the serialize isn't very symmetrical with the deserialize method, but I feel like they *should* be (one should just be doing the opposite of the other.

kellyredding commented on lib/qs/payload.rb:
```diff
@@ -0,0 +1,47 @@
+        :type       => payload_hash['type'],
+        :created_at => Time.at(payload_hash['created_at'].to_i)
+      })
+    end
```
@jcredding I don't love that the method name describes what it takes, not what it returns.

kellyredding commented on lib/qs/payload.rb:
```diff
@@ -0,0 +1,47 @@
+
+    def self.serialize_hash(payload_hash)
+      Qs.serialize(payload_hash)
+    end
```
@jcredding also don't love that this method name describes what it takes, not what it returns.

Also, this method is nothing more than an alias for `Qs.serialize` which seems like little value add.

kellyredding commented:
@jcredding Ok, that is it.  Would love your thoughts.  The cool thing is that even this implementation is enough to justify this new module of functions.  I love the removals on the job object.

jcredding commented:
@kellyredding - Overall I don't have a problem with what you are proposing and I'll work up the changes and we can discuss further. 

* I was going to add the `job` method later once this handled both jobs and events (using `send`) so no issue with that. 
* The `serialize_hash` method is essentially an alias because I was trying to run everything through `Qs::Payload` and not call `Qs.serialize` directly. This was an attempt to reduce coupling. The difference is the clients will need both `Qs.encode` (for `push`) and `Qs::Payload.serialize`. Not a big deal.
* We may need to rename the `hash` method to `job_hash` once we are handling events. I've started the event inheriting from job stuff and there's some weirdness with the event info being buried in the params. There's a long explanation for the why, but the core of the issue is we want `Event#params` to return the full params in (with channel, name and params buried in it) when the dispatcher builds an event but we want the params method to just return the event params for the event handler. There's ways around this and I think the base `Message` class might actually be the cleanest way (I've started looking at this). Anyway, this would also keep the methods even more consistent:

```ruby
def self.deserialize(serialized_payload)
  payload = Qs.decode(serialized_payload)
  self.send(payload['type'], payload)
end

def self.serialize(job)
  payload = self.send("#{job.payload_type}_hash", job)
  Qs.encode(payload)
end
```

Good stuff, thanks for the thorough explanation.

kellyredding commented:
@jcredding oooh, I like the serialize/deserialize implementations you proposed and how they use the the payload type to build the right stuff.  I like how they are really consistent.

The more I'm thinking about Qs, I *really* wanna see a base `Message` class come about.  Then (hopefully) `serialize` would just take a `message` arg:

```ruby
def self.serialize(message)
  payload = self.send("#{message.payload_type}_hash", message)
  Qs.encode(payload)
end
```

jcredding commented:
@kellyredding - I've updated this like we discussed. I renamed serialize/deserialize to encode/decode and changed the `Payload` module methods. I'm happy with the changes and I agree, I like the idea of the base `Message` class.

kellyredding commented:
@jcredding this all looks really good.  Nice reworks :-1:
